### ISSUE TITLE
AmazonSimpleNotificationService PublishAsync intermittently getting request signature we calculated does not match the signature you provided.

### ISSUE THREAD

**Issue by Laurence-yau** [aws/aws-sdk-net#2545]
### Describe the bug

AmazonSimpleNotificationService PublishAsync intermittently gets an error  "The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details"

We have tried triggering the code path but we have been unable to reproduce this error. The service that gets the error publishes many other SNS topics however it only happens with two specific SNS the service sends, all other sns we send have not gotten this error. 

### Expected Behavior

The SNS to be sent .

### Current Behavior

``Amazon.SimpleNotificationService.AmazonSimpleNotificationServiceException: The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.
 ---> Amazon.Runtime.Internal.HttpErrorResponseException: Exception of type 'Amazon.Runtime.Internal.HttpErrorResponseException' was thrown.
   at Amazon.Runtime.HttpWebRequestMessage.GetResponseAsync(CancellationToken cancellationToken)
   at Amazon.Runtime.Internal.HttpHandler`1.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.Unmarshaller.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.ErrorHandler.InvokeAsync[T](IExecutionContext executionContext)
   --- End of inner exception stack trace ---
   at Amazon.Runtime.Internal.HttpErrorResponseExceptionHandler.HandleExceptionStream(IRequestContext requestContext, IWebResponseData httpErrorResponse, HttpErrorResponseException exception, Stream responseStream)
   at Amazon.Runtime.Internal.HttpErrorResponseExceptionHandler.HandleExceptionAsync(IExecutionContext executionContext, HttpErrorResponseException exception)
   at Amazon.Runtime.Internal.ExceptionHandler`1.HandleAsync(IExecutionContext executionContext, Exception exception)
   at Amazon.Runtime.Internal.ErrorHandler.ProcessExceptionAsync(IExecutionContext executionContext, Exception exception)
   at Amazon.Runtime.Internal.ErrorHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.Signer.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CredentialsRetriever.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.MetricsHandler.InvokeAsync[T](IExecutionContext executionContext)``

### Reproduction Steps

We have been unable to reproduce the behaviour. 

```
           var tasks = new List<Task<PublishResponse>>();
            var messages = new List<Message>();

            foreach (var message in messages)
            {
                tasks.Add(_snsClient.PublishAsync(new PublishRequest
                {
                    TopicArn = message.SnsTopicArn,
                    Message = message.Payload,
                    MessageAttributes = message.MessageAttributes,

                }));
            }

            await Task.WhenAll(tasks);
```

### AWS .NET SDK and/or Package version used

AWSSDK.SimpleNotificationService 3.7.2.59

### Targeted .NET Platform

.Net 6

### Operating System and version

Ubuntu 20.04 LTS

**Comment by ashishdhingra** [comment]
Hi @Laurence-yau,

Good morning.

Thanks for reporting the issue. Could you please share the following:
- How are the credentials for SNS client configured/resolved?
- Try enabling the verbose logging on the affected SNS endpoints and share the verbose logs. Verbose logging could be enabled using below code as an example:
```C#
Amazon.AWSConfigs.LoggingConfig.LogResponses = Amazon.ResponseLoggingOption.Always;
Amazon.AWSConfigs.LoggingConfig.LogTo = Amazon.LoggingOptions.SystemDiagnostics;
Amazon.AWSConfigs.AddTraceListener("Amazon", new System.Diagnostics.ConsoleTraceListener());
```
- What is the average number of messages processed? When error occurs, how many messages were processed?
- Try limiting the amount of messages that gets processed in your code. In other words, consider the optimal batch size (based on testing), lets say 10 and process that many SNS messages in a batch. That might prevent throttling and resolve such issue.

Thanks,
Ashish

**Comment by Laurence-yau** [comment]
> Hi @Laurence-yau,
> 
> Good morning.
> 
> Thanks for reporting the issue. Could you please share the following:
> 
> * How are the credentials for SNS client configured/resolved?
> * Try enabling the verbose logging on the affected SNS endpoints and share the verbose logs. Verbose logging could be enabled using below code as an example:
> 
> ```cs
> Amazon.AWSConfigs.LoggingConfig.LogResponses = Amazon.ResponseLoggingOption.Always;
> Amazon.AWSConfigs.LoggingConfig.LogTo = Amazon.LoggingOptions.SystemDiagnostics;
> Amazon.AWSConfigs.AddTraceListener("Amazon", new System.Diagnostics.ConsoleTraceListener());
> ```
> 
> * What is the average number of messages processed? When error occurs, how many messages were processed?
> * Try limiting the amount of messages that gets processed in your code. In other words, consider the optimal batch size (based on testing), lets say 10 and process that many SNS messages in a batch. That might prevent throttling and resolve such issue.
> 
> Thanks, Ashish

Hi Ashish,

1.  In the task definition for our containers we specify the taskRoleArn. For the SDK as we are using the AWSSDK.Extensions.NETCore.Setup and don't specify credentials. Based on what I understand it should be retrieving the credentials from AWS_CONTAINER_CREDENTIALS_RELATIVE_URI. 
2. Will give this a go
3. The code block will process between 1-3 messages. Nothing substantial, that should be causing us to get throttled. The behaviour we are seeing is the error occurs when there is more than 2 messages.

Thanks,
Laurence

**Comment by Laurence-yau** [comment]
Hi Ashish,

On further investigation of our logs, this intermittent issue may not be isolated to SNS. We seems to be having a similar with SQS. See below

```
 Amazon.SQS.AmazonSQSException: The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.

The Canonical String for this request should have been
'POST
/{accountNumber}/{queueName}

content-type:application/x-www-form-urlencoded; charset=utf-8
host:sqs.ap-southeast-2.amazonaws.com
user-agent:aws-sdk-dotnet-coreclr/3.7.1.31 aws-sdk-dotnet-core/3.7.103.2 .NET_Core/6.0.13 OS/Linux_4.14.304-226.531.amzn2.x86_64_#1_SMP_Wed_Feb_1_21:34:38_UTC_2023 exec-env/AWS_ECS_EC2 ClientAsync aws-sdk-dotnet-coreclr/3.7.1.31 aws-sdk-dotnet-core/3.7.103.2 .NET_Core/6.0.13 OS/Linux_4.14.304-226.531.amzn2.x86_64_#1_SMP_Wed_Feb_1_21:34:38_UTC_2023 exec-env/AWS_ECS_EC2 ClientAsync
x-amz-content-sha256:fbca1e932d021e133163b5cf1ab418d4526e7320f499a23cxxxx9831a7ad75e4
x-amz-date:20230222T195838Z
x-amz-security-token:{token}

content-type;host;user-agent;x-amz-content-sha256;x-amz-date;x-amz-security-token
fbca1e932d021e133163b5cf1ab418d4526e7320f499a23cxxxx9831a7ad75e4'

The String-to-Sign should have been
'AWS4-HMAC-SHA256
20230222T195838Z
20230222/ap-southeast-2/sqs/aws4_request
0bd9d85e5db2f435c866b776b8a42d5592c8edcc50fdfade695801d9bb53xxxx'

 ---> Amazon.Runtime.Internal.HttpErrorResponseException: Exception of type 'Amazon.Runtime.Internal.HttpErrorResponseException' was thrown.
   at Amazon.Runtime.HttpWebRequestMessage.GetResponseAsync(CancellationToken cancellationToken)
   at Amazon.Runtime.Internal.HttpHandler`1.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.Unmarshaller.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.SQS.Internal.ValidationResponseHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.ErrorHandler.InvokeAsync[T](IExecutionContext executionContext)
   --- End of inner exception stack trace ---
   at Amazon.Runtime.Internal.HttpErrorResponseExceptionHandler.HandleExceptionStream(IRequestContext requestContext, IWebResponseData httpErrorResponse, HttpErrorResponseException exception, Stream responseStream)
   at Amazon.Runtime.Internal.HttpErrorResponseExceptionHandler.HandleExceptionAsync(IExecutionContext executionContext, HttpErrorResponseException exception)
   at Amazon.Runtime.Internal.ExceptionHandler`1.HandleAsync(IExecutionContext executionContext, Exception exception)
   at Amazon.Runtime.Internal.ErrorHandler.ProcessExceptionAsync(IExecutionContext executionContext, Exception exception)
   at Amazon.Runtime.Internal.ErrorHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.Signer.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CredentialsRetriever.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.MetricsHandler.InvokeAsync[T](IExecutionContext executionContext)
   at MassTransit.AmazonSqsTransport.Contexts.AmazonSqsClientContext.ChangeMessageVisibility(String queueUrl, String receiptHandle, Int32 seconds)
   at MassTransit.AmazonSqsTransport.Contexts.AmazonSqsReceiveContext.Faulted(Exception exception)
```

**Comment by ashishdhingra** [comment]
@Laurence-yau The part that was missing in the issue description earlier was the fact that the application is running inside the ECS container (since you mentioned about `AWS_CONTAINER_CREDENTIALS_RELATIVE_URI`). Although the credentials should refresh automatically, my hunch is that when the credentials are actually used to make the service call, these are expired, also given that the issue is intermittent. There is also a similar issue https://github.com/aws/aws-sdk-net/issues/2498 in our backlog.

Needs review with the team.

**Comment by Laurence-yau** [comment]
Hi Ashish,

Whilst I definitely agree that the above issue is happening in our system. I am a bit reluctant to write the specific SNS issue as being part of that just yet.

I have nothing concrete at the moment and am still investigating though. But we had no such occurrence of the error in our production environment until our last release, in the last release we added an extra message in the code path so that it could potentially be publish up to 3 messages, which seems to be the tipping point.

We have changed the code to wait for each publish instead of firing off all of them at once now. So far we have yet to see any occurrence of the issue again.

```
            var messages = new List<Message>();

            foreach (var message in messages)
            {
                await _snsClient.PublishAsync(new PublishRequest
                {
                    TopicArn = message.SnsTopicArn,
                    Message = message.Payload,
                    MessageAttributes = message.MessageAttributes,

                });
            }

```



Thanks,
Laurence

**Comment by Laurence-yau** [comment]
I can confirm that the code change stops the issue from happening. We haven't seen the issue since.

**Comment by peterrsongg** [comment]
The fix for this was released in 3.7.506.0. We believe it was an ECS task credential refresh issue. If the issue still persists with the original code then feel free to re-open this issue.

**Comment by gregbradley** [comment]
> The fix for this was released in 3.7.506.0. We believe it was an ECS task credential refresh issue. If the issue still persists with the original code then feel free to re-open this issue.

Do you mean version 3.7.106.0?

**Comment by peterrsongg** [comment]
@gregbradley Yes in Core version 3.7.106.0 from (2023-03-17 13:51 UTC)
